### ISSUE TITLE
xmlgen: Incorrect codegen with more complex type structures

### ISSUE THREAD

**Issue by zeenix** [dbus2/zbus#257]
In GitLab by @PolyMeilex on Mar 31, 2022, 12:20

Hi, I have a method like this one:
```xml
    <method name="GetAbcd">
      <arg type="u" name="a" direction="out" />
      <arg type="a((ssss)a(siiddada{sv})a{sv})" name="b" direction="out" />
      <arg type="a(iiduba(ssss)a{sv})" name="c" direction="out" />
      <arg type="a{sv}" name="d" direction="out" />
    </method>
```
After piping it into xmlgen I get this type as return value:
```rs
     (
        u32,
        Vec<(String, String, String, String)>,
        Vec<(
            i32,
            i32,
            f64,
            u32,
            bool,
            Vec<(String, String, String, String)>,
        )>,
        std::collections::HashMap<String, zbus::zvariant::OwnedValue>,
    )

```
There is several fields missing in this definition, here is the same type manually written by me to match what xml defines:
```rs
(
    u32,
    Vec<(
        (String, String, String, String),
        Vec<(
            String,
            i32,
            i32,
            f64,
            f64,
            Vec<f64>,
            HashMap<String, zbus::zvariant::OwnedValue>,
        )>,
        HashMap<String, zbus::zvariant::OwnedValue>,
    )>,
    Vec<(
        i32,
        i32,
        f64,
        u32,
        bool,
        Vec<(String, String, String, String)>,
        HashMap<String, zbus::zvariant::OwnedValue>,
    )>,
    HashMap<String, zbus::zvariant::OwnedValue>,
)
```

**Comment by nagatavit** [comment]
Hi, I stumbled on this problem today, and dig a little to find out if it was easy to fix, and it was! I've opened the PR #506, however, I never contributed to an open source project, so I'm kinda lost on what I should do, next.

Can someone give me some pointers for the next steps? :thinking:

**Comment by zeenix** [comment]
> Hi, I stumbled on this problem today, and dig a little to find out if it was easy to fix, and it was! I've opened the PR #506,

Thanks so much. :pray: 

> however, I never contributed to an open source project, so I'm kinda lost on what I should do next.

Really? You did great and even followed the contributors guide to the letter. Most folks don't do it so well.

> Can someone give me some pointers on how to proceed? 🤔

What happens next is that I (and/or others) review the PR, possibly ask for changes before it can be merged. Once the changes are good and the CI passes (I've to approve the CI run manually for first contributors), I'll merge your PR.

Unless you wanted to contribute more? In which case, I've a [bunch of issues](https://github.com/dbus2/zbus/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) that could be good for new contributors to pick up.

**Comment by nagatavit** [comment]
> > Hi, I stumbled on this problem today, and dig a little to find out if it was easy to fix, and it was! I've opened the PR #506,
> 
> Thanks so much. 🙏 
> > however, I never contributed to an open source project, so I'm kinda lost on what I should do next.
> 
> Really? You did great and even followed the contributors guide to the letter. Most folks don't do it so well.

Thanks! Yeah, I didn't contributed to any open source before (other than for some friends projects), but I have some notions of looking for standards and rules.

> > Can someone give me some pointers on how to proceed? 🤔
> 
> What happens next is that I (and/or others) review the PR, possibly ask for changes before it can be merged. Once the changes are good and the CI passes (I've to approve the CI run manually for first contributors), I'll merge your PR.

:+1: 

> Unless you wanted to contribute more? In which case, I've a [bunch of issues](https://github.com/dbus2/zbus/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) that could be good for new contributors to pick up.

Good to know! If I have the time I'll try to look into it!
